File: src/main/java/juuxel/woodsandmires/feature/FellLakeFeature.java
package juuxel.woodsandmires.feature;

import com.mojang.serialization.Codec;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.world.gen.feature.LakeFeature;
import net.minecraft.world.gen.feature.SingleStateFeatureConfig;

import java.util.Random;

public class FellLakeFeature extends LakeFeature implements LakeFeatureExtensions {
    public FellLakeFeature(Codec<SingleStateFeatureConfig> codec) {
        super(codec);
    }

    @Override
    public boolean wam_shouldReplaceBorderingBlocks() {
        return true;
    }

    @Override
    public BlockState wam_getBorderBlock(Random random) {
        if (random.nextInt(60) == 0) {
            return Blocks.EMERALD_ORE.getDefaultState();
        } else if (random.nextInt(55) == 0) {
            return Blocks.GOLD_ORE.getDefaultState();
        }

        return Blocks.STONE.getDefaultState();
    }
}

File: src/main/kotlin/juuxel/woodsandmires/feature/FellLakeFeature.kt
package juuxel.woodsandmires.feature

import com.mojang.serialization.Codec
import net.minecraft.block.Blocks
import net.minecraft.world.gen.feature.LakeFeature
import net.minecraft.world.gen.feature.SingleStateFeatureConfig
import java.util.Random

class FellLakeFeature(
    configCodec: Codec<SingleStateFeatureConfig>
) : LakeFeature(configCodec), LakeFeatureExtensions {
    override fun wam_shouldReplaceBorderingBlocks() = true

    override fun wam_getBorderBlock(random: Random) =
        when {
            random.nextInt(60) == 0 -> Blocks.EMERALD_ORE.defaultState
            random.nextInt(55) == 0 -> Blocks.GOLD_ORE.defaultState
            else -> Blocks.STONE.defaultState
        }
}
